get-terraform-provider-versions/get_terraform_provider_versions/get-tf-provider-versions.py

```diff
@@ -0,0 +1,337 @@
+#!/usr/bin/env python3
+import sys
+import getopt
+import os
+import re
+import logging
+import tempfile
+import subprocess
+import requests
+
+from github.repo import Repo
+
+GITHUB_ORGANIZATION = 'dfds'
+TFREGISTRY_BASEAPI = 'https://registry.terraform.io/v1/providers/'
+
+
+# define a custom class to hold the providers we locate
+class TerraformProvider:
+    """
+    This class defines a Terraform provider
+    """
+
+    def __init__(self, repository_name: str, file_path: str, provider_name: str, provider_version: str, latest_provider_version: str = "", comment: str = ""):
+        self.repository_name = repository_name
+        self.file_path = file_path
+        self.provider_name = provider_name
+        self.provider_version = provider_version
+        self.latest_provider_version = latest_provider_version
+        self.comment = comment
+
+
+class TerraformProviders:
+    """
+    This class defines a collection of Terraform providers
+    """
+
+    def __init__(self):
+        self.terraform_providers: list = []
+
+    def __iter__(self):
+        return iter(self.terraform_providers)
+
+    def append(self, new_item):
+        self.terraform_providers.append(new_item)
+
+    def count(self):
+        return len(self.terraform_providers)
+
+    def get_json(self):
+        """
+        Get the data from the collection elements and return them in JSON format.
+        """
+        json_string: str = '['
```

Comment:
The easiest way to work with JSON in Python is through the use of dictionaries (dict). 
Other languages might have collections of similar functionalities. It is basically a map of key/value pairs.
```

my_list: list = []
for provider in self.terraform_providers:
    my_dict: dict = {'repository_name': provider.repository_name,
                     'terraform_file': provider.file_path,
                     and so on...
                     }
    my_list.append(my_dict)
return json.dumps(my_list)
```

Reply:
Ok, this was quite helpful but it required a bit more work for me to get it working.

TerraformProvider is a custom class that i've built and the data is stored as a series of parameters.  However, I found that you could cast it to a dictionary type by using the __dict__ method.  So the code now looks like this.  It's less complex and also we don't need to handle escaping the backslashes as it's taken care for us.

```
        json_string: str = "["
        for provider in self.terraform_providers:
            json_string += (json.dumps(provider.__dict__)) + ","
        json_string = json_string.strip(",") + "]"
        return json_string
```